sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,642 @@
+can all be uniquely referenced via fully qualified lookup (e.g.
+`::inner::Bar`).  However, it is possible for users to construct applications
+where this is not the case, for example:
+
+```
+sycl::device_global<int> FuBar;
+namespace {
+  sycl::device_global<int> FuBar;
+}
+```
+
+In this example, the `FuBar` variable in the global namespace shadows a
+variable with the same name in the unnamed namespace.  The integration footer
+can reference the variable in the global namespace as `::FuBar`, but there is
+no way to reference the variable in the unnamed namespace using fully qualified
+lookup.
+
+Such programs are still legal, though.  The integration footer can support
+cases like this by defining a temporary variable that holds the address of the
+shadowed device global:
+
+```
+namespace {
+const void *__sycl_UNIQUE_STRING = &FuBar;  // References 'FuBar' in the
+                                            // unnamed namespace
+}
+
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  device_global_map::add(&::FuBar,
+    /* same string returned from __builtin_sycl_unique_stable_id(::FuBar) */);
+  device_global_map::add(::__sycl_UNIQUE_STRING,
+    /* same string returned from __builtin_sycl_unique_stable_id(::(unnamed)::FuBar) */);
+}
+
+} // namespace (unnamed)
+} // namespace sycl::detail
+```
+
+The `__sycl_UNIQUE_STRING` variable is defined in the same namespace as the
+second `FuBar` device global, so it can reference the variable through
+unqualified name lookup.  Furthermore, the name of the temporary variable
+(`__sycl_UNIQUE_STRING`) is globally unique, so it is guaranteed not to be
+shadowed by any other name in the translation unit.  This problem with variable
+shadowing is also a problem for the integration footer we use for
+specialization constants.  See the [specialization constant design document][5]
+for more details on this topic.
+
+[5]: <SpecializationConstants.md>
+
+### Changes to the `sycl-post-link` tool
+
+The `sycl-post-link` tool performs its normal algorithm to identify the set of
+kernels and device functions that are bundled together into each module.  Once
+it identifies the functions in each module, it scans those functions looking
+for references to global variables that are decorated with the LLVM IR
+attribute `"sycl-device-global-size"` (these are the variables of type
+`device_global`).  If any device global variable decorated with the LLVM IR
+attribute corresponding to the `device_image_scope` property appears in more
+than one module, the `sycl-post-link` tool issues an error diagnostic:
+
+```
+error: device_global variable <name> with property "device_image_scope"
+       is contained in more than one device image.
+```
+
+Assuming that no error diagnostic is issued, the `sycl-post-link` tool includes
+the IR definition of each `device_global` variable in the modules that
+reference that variable.
```

Comment:
@gmlueck, I've just realized that we might want to be more precise here. In particular, at the moment `sycl-post-link` clones **all** global variables into every produced device image:

https://github.com/intel/llvm/blob/7ae7ca872ffcfdcce6d6f23ce8260705e1250241/llvm/tools/sycl-post-link/sycl-post-link.cpp#L480-L485

I assume that could be a problem, if we later produce device image properties with unused globals. At least we will waste some memory and probably compute time. What do you think?

Reply:
It should be correct to include all global variables in all modules.  As you say, however, this will be wasteful.  For example, consider some very large device global that is referenced by only one kernel.  It would waste device memory to include this variable in all modules.  This might be particularly problematic on FPGA, where device memory is more limited.

What about the current wording in the design doc should be made more precise?  That wording already says that the IR for a device global should only be copied to a module if the module references the variable.

Reply:
> What about the current wording in the design doc should be made more precise? That wording already says that the IR for a device global should only be copied to a module if the module references the variable.

Maybe I'm misreading it, but I don't seem to find "should _**only**_ be copied" part, i.e. I don't see the part which restricts `sycl-post-link` from doing extra copies.

> It should be correct to include all global variables in all modules. As you say, however, this will be wasteful. For example, consider some very large device global that is referenced by only one kernel. It would waste device memory to include this variable in all modules. This might be particularly problematic on FPGA, where device memory is more limited.

Ok, thanks for confirming. I think this conversation can be marked as resolved